docs: modify hermetic build docs (#3331)
googleapis/sdk-platform-java: hermetic_build/README.md

> [!IMPORTANT]
> All scripts/examples assume you are inside the repository root folder.

# Generate a repository containing GAPIC Client Libraries

Running the docker image built from `hermetic_build/library_generation`
directory, you can generate a repository containing GAPIC client libraries (a
monorepo, for example, google-cloud-java) from a configuration file.

Instead of running docker image, if you prefer running the underlying python

> script directly, please refer to [development guide](DEVELOPMENT.md#run-the-script) for

Author: Done.

> additional instructions.

Author: I think the purpose of README is to instruct users how to do things in the most convenient way, thus I moved the part of running the script directly to development guide as it requires additional setups that users typically don't bother to do.

Instead, I moved the image building part here because it's the most straightforward way (i.e., no need to install dependencies) of generating a repository.

Reviewer: I think the differences between README and DEVELOPMENT is that, README is for external users that just want to run our hermetic image as a product, while DEVELOPMENT is for internal developers that try to maintain the scripts and images. 
Hence for README users, they can think hermetic code generation as a blackbox, all they need to know is the input and output. For DEVELOPMENT users, they need to understand how the process uses Python/Shell/Docker etc, so build image locally is better suited for DEVELOPMENT.

Author: SG, I moved the instructions on building image from source to development guide.

## Environment

- OS: Linux
- Docker

## Prerequisite

In order to generate a version for each library, a versions.txt has to exist
in `repository-path`.
Please refer to [Repository path](#repository-path--repositorypath---optional)
for more information.

## Parameters to generate a repository using the docker image

### Generation configuration yaml (`generation-config-path`)

An absolute or relative path to a configuration file containing parameters to
generate the repository.

> Please refer [Configuration to generate a repository](#configuration-to-generate-a-repository)

Author: Done.

Please refer [here](go/java-client-releasing#versionstxt-manifest) for more info
of versions.txt.

### A list of library names (`library-names`), optional

> A list of library names that will be generated, separated by comma.

Reviewer 2: will this be written to stdout or a file?

Author: This is one of the input.

The library name of a library is the value of `library_name` or `api_shortname`,
if `library_name` is not specified, in the generation configuration.

If not specified, all libraries in the generation
configuration will be generated.

### Api definitions path (`api-definitions-path`), optional

The path to where the api definition (proto, service yaml) resides.

The default value is the current working directory when running the script.

> Note that you need not only the protos defined the service, but also the

Reviewer: I think this paragraph deserve to be bolded.

Author: Done.

| distribution_name     |    No    | `{group_id}:google-{cloud_prefix}{library_name}` if not specified                                                                         |
| excluded_poms         |    No    |                                                                                                                                           |
| excluded_dependencies |    No    |                                                                                                                                           |

> | googleapis_commitish  |    No    | use repository level `googleapis_committish` if not specified.                                                                            |

Reviewer 2: ```suggestion
| googleapis_commitish  |    No    | use repository level `googleapis_committish` if not specified.                                                                            |
```
Not the right spelling but it's what's defined in the repo-level section.

Author: Yes, it should be `googleapis_committish` according to https://github.com/googleapis/sdk-platform-java/pull/3208#discussion_r1765691116.

However, I think the refactor should be in its own PR.

Reviewer 2: Sorry I misspelled the suggestion string. I was trying to point out that we should stick to `googleapis_commitish` (one `t`) since we haven't done such refactor.

      - proto_path: google/cloud/asset/v1p7beta1
```

# Run the library generation image

1. Download the API definitions to a local directory, e.g., from [googleapis](https://github.com/googleapis/googleapis).

2. Run the docker image.
   ```shell
   # Assume you want to generate the library in the current working directory
   # and the generation configuration is in the same directory.
   docker run \
     --rm \
     --quiet \
     -u "$(id -u):$(id -g)" \
     -v "$(pwd):/workspace" \
     -v /path/to/api_definition:/workspace \

>      -e GENERATOR_VERSION=image-tag \

Reviewer: Do we have to set `GENERATOR_VERSION`? Is it still used in our scritps? Can we just inline the image-tag?

Author: OK, I can see generator version is defined in the generation config example.

I'll remove this param.

      - proto_path: google/cloud/asset/v1p7beta1
```

# Run the library generation image

1. Download the API definitions to a local directory, e.g., from [googleapis](https://github.com/googleapis/googleapis).

2. Run the docker image.
   ```shell
   # Assume you want to generate the library in the current working directory
   # and the generation configuration is in the same directory.

>    docker run \

Reviewer: This command includes all the possible parameters, can we provide an example with minimum parameters?

Author: Done.

Reviewer: Your original example still makes sense for more advanced use cases, I didn't mean to ask you to delete it, we can still have another example with all the parameters. On the other hand, maybe we don't have to provide a more complex example as it could confuse other users. So either way works for me, feel free to decide as you see fit.

Author: I added back the advance example with explanations.

      - proto_path: google/cloud/asset/v1p7beta1
```

# Build the image from source

1. Run the following command to build the image from source

   ```shell
   docker build \
     -f .cloudbuild/library_generation/library_generation.Dockerfile \
     -t local:image-tag \
     .
   ```

> 2. Download api definition to a local directory

Reviewer 2: ```suggestion
2. Download the API definitions to a local directory
```

Should this offer an example of where the API definitions are located?

|_versions.txt
```

> # Generate release note from API definition changes

Reviewer 2: ```suggestion
# Generate release notes from API definition changes
```

Author: Done.

|_versions.txt
```

# Generate release note from api definition changes

The script, `hermetic_build/release_note_generation/cli/generate_release_note.py`

> allows you to generate release note from api definition changes in

Reviewer 2: ```suggestion
allows you to generate a release notes file from the API definition changes in the
```

[googleapis](https://github.com/googleapis/googleapis) GitHub repository.

## Environment

- OS: Linux
- Python (3.12.0 or above)

## Parameters to generate a release note

### Baseline generation configuration path (`baseline-generation-config-path`)

Absolute or relative path to a generation configuration.

> Please refer [Configuration to generate a repository](#configuration-to-generate-a-repository)

Reviewer 2: ```suggestion
Please refer to [Configuration to generate a repository](#configuration-to-generate-a-repository)
```

Author: Done.

for more information.

Note that the `googleapis_commitish` in this configuration is used to retrieve

> the first commit, exclusively, to generate the release note.

Reviewer 2: ```suggestion
the first commit, exclusively, to generate the release notes.
```

Author: Done.

### Current generation configuration path (`current-generation-config-path`)

Absolute or relative path to a generation configuration.

> The release note will be generated from commits that related to the libraries

Reviewer 2: ```suggestion
The release notes will be generated from commits that are related to the libraries
```

Author: Done.

specified in this configuration.

> Please refer [Configuration to generate a repository](#configuration-to-generate-a-repository)

Reviewer 2: ```suggestion
Please refer to [Configuration to generate a repository](#configuration-to-generate-a-repository)
```

Author: Done.

for more information.

> Note that the `googleapis_commitish` in this configuration is used to retrieve
> the last commit, inclusively, to generate the release note.

Reviewer 2: ```suggestion
Note that the `googleapis_commitish` entry in this configuration is used to retrieve
the last commit, inclusively, to generate the release notes.
```

### Repository path (`repository-path`), optional

> The path to which the file, `pr_description.txt` containing the release note

Reviewer 2: ```suggestion
The path to which the file, `pr_description.txt` containing the release notes
```

will be sent.
If not specified, the file will be generated to the current working directory.

> ## Generate a release note in local environment

Reviewer 2: ```suggestion
## Generating a release notes file in a local environment
```

Author: Done.

> 1. Install python (>= 3.12.0), it is recommended to create a python virtual
> environment through [official guide](https://packaging.python.org/en/latest/guides/installing-using-pip-and-virtual-environments/#create-and-use-virtual-environments).

Reviewer 2: ```suggestion
1. Install python (>= 3.12.0). It is recommended to create a python virtual
environment through the [official guide](https://packaging.python.org/en/latest/guides/installing-using-pip-and-virtual-environments/#create-and-use-virtual-environments).
```

Author: Done.

> 2. Running the following commands to install python packages

Reviewer 2: ```suggestion
2. Run the following commands to install python packages
```

   ```shell
   cd /path/to/sdk-platform-java 
   pip install --require-hashes -r hermetic_build/common/requirements.txt
   pip install hermetic_build/common
   pip install --require-hashes -r hermetic_build/release_note_generation/requirements.txt
   pip install hermetic_build/release_note_generation
   ```

> 3. Running the following commands to generate a release note

Reviewer 2: ```suggestion
3. Run the following commands to generate a release note
```

Author: Done.
